ibis/expr/operations/udf.py

```diff
@@ -0,0 +1,149 @@
+from __future__ import annotations
+
+import functools
+import inspect
+import typing
+from typing import TYPE_CHECKING, Any, Callable, TypeVar
+
+from public import public
+
+import ibis.common.exceptions as exc
+import ibis.expr.datatypes as dt
+import ibis.expr.operations as ops
+import ibis.expr.rules as rlz
+from ibis import util
+from ibis.common.collections import FrozenDict
+
+if TYPE_CHECKING:
+    import ibis.expr.types as ir
+
+
+@public
+class ScalarUDF(ops.Value):
+    output_shape = rlz.shape_like("args")
+
+
+@public
+class SingleScalarUDF(ScalarUDF):
+    pass
+
+
+@public
+class VectorizedScalarUDF(ScalarUDF):
+    pass
+
+
+def _wrap(wrapper, fn: Callable | None = None, *args: Any, **kwargs: Any) -> Callable:
+    """Wrap a function `fn` with `wrapper`, allowing zero arguments when used as part of a decorator."""
+    if fn is None:
+        return lambda fn: functools.update_wrapper(wrapper(fn, *args, **kwargs), fn)
+    return functools.update_wrapper(wrapper(fn, *args, **kwargs), fn)
+
+
+S1 = TypeVar("S1", bound=ScalarUDF)
+S2 = TypeVar("S2", bound=ScalarUDF)
+
+
+class UDF:
+    """Construct wrappers for user-defined functions."""
+
+    @util.experimental
+    def __call__(
+        self, fn: Callable | None = None, *args: Any, **kwargs: Any
+    ) -> Callable:
+        """Construct a scalar user-defined function.
+
+        Examples
+        --------
+        >>> from ibis.interactive import *
+        >>> @udf.scalar
```

Comment:
I would prefer this to be explicitly elementwise, I don't like the bare @udf.scalar

@udf.scalar.elementwise
and
@udf.scalar.vectorized

of course I could see someone wanting .vectorized to either use pandas or pyarrow, so following this pattern would allow for extension, eg
- @udf.scalar.elementwise
- @udf.scalar.pandas
- @udf.scalar.arrow

Reply:
Not opposed to `@udf.scalar.python` and the others. I would like to avoid too much `.`-ing since it can get confusing to know in what order things should be spelled when there are a lot of dots necessary.

Reply:
I was able to implement these:

```
@udf.scalar.python
@udf.scalar.pandas
@udf.scalar.pyarrow
```